Added built-in prompt moderation for OpenAI endpoints

YashKarthik opened the pull request:
- Checks if `Helicone-Moderations-Enabled: true` before hitting moderation endpoint.
- Only runs for OPENAI provider.
- Only runs when role == "user" for message.
- Uses the generic `handleProxyRequest` function to call moderations endpoint. (builds proxyRequest object for the mods call).

chitalian commented on worker/package-lock.json:
Can you remove this file please?

chitalian commented on valhalla/jawn/package-lock.json:
Can you remove this file please?
...
We should add these files to the .gitignore for anyone using npm

chitalian commented on worker/src/lib/HeliconeProxyRequest/forwarder.ts:
```diff
@@ -130,44 +173,98 @@ export async function proxyForwarder(
-            )
-          );
-          return cachedResponse;
+    if (latestMessage.role == "user") {
```
Last message has an implicit type of any.

can you change the type of `latestMessage` to be `myType | null` and set it to null above to safely grab content?

```ts
let latestMessage: {
   user?: string,
   content?: unknown
} | null = null;
```
```suggestion
if (latestMessage !== null && latestMessage?.role == "user")
````

YashKarthik replied:
Hi, any reason to have `content?: unknown` instead of `content?: string`?

chitalian commented on worker/src/lib/HeliconeProxyRequest/forwarder.ts:
```diff
@@ -130,44 +173,98 @@ export async function proxyForwarder(
+          method: "POST",
+          headers: proxyRequest.requestWrapper.headers,
+          body: JSON.stringify({
+            input: latestMessage.content,
```
content might not exist on lastMessage

YashKarthik replied:
Should we return an error when the content is empty/does not exist?

colegottdank replied:
IMO, if no content, we just don't check moderation and continue. Maybe there is a reason and we want the provider to fail, not us. 

I would add to the latestMessage?.role if check if the content exists

chitalian commented:
Great work @YashKarthik !! :D

chitalian commented on worker/src/lib/HeliconeProxyRequest/forwarder.ts:
```diff
@@ -130,44 +173,98 @@ export async function proxyForwarder(
+        });
+      }
+
+      const { data: moderationResponse, error: moderationResponseError } =
+        await handleProxyRequest(moderationProxyRequest);
+      if (moderationResponseError != null) {
+        return responseBuilder.build({
+          body: moderationResponseError,
+          status: 500,
+        });
+      }
+
+      const flaggedForModeration = (
+        (await moderationResponse.response.json()) as OpenAIModerationResponse
+      ).results[0].flagged;
+      proxyRequest.flaggedForModeration = flaggedForModeration;
+      ctx.waitUntil(log(moderationResponse.loggable));
+
+      if (flaggedForModeration == true) {
+        moderationResponse.response.headers.forEach((value, key) => {
+          responseBuilder.setHeader(key, value);
+        });
+
+        const responseContent = {
+          body: JSON.stringify({
+            success: false,
+            error: {
+              code: "PROMPT_FLAGGED_FOR_MODERATION",
+              message:
+                "The given prompt was flagged by the OpenAI Moderation endpoint.",
+              details: `See your Helicone request page for more info: https://www.helicone.ai/requests?${moderationProxyRequest.requestId}`,
+            },
+          }),
+          inheritFrom: moderationResponse.response,
+          status: 400,
+        };
+
+        return responseBuilder
+          .setHeader("content-type", "application/json")
+          .build(responseContent);
```
This is a lot of code to add within this method. @colegottdank thoughts on moving this to another file? 

Maybe we can create a moderator class and then we can do something like.


```ts
const mod = new Moderator(...my vars)

const moderationResult = mod.moderate();

if (moderationResult.moderated.isModerated === true) {
    return moderationResult.moderationResponse;
}
```

YashKarthik commented:
Hi, some changes:
- improved the error/null handling around `latestMessage`. Returns a 400 to user if the content is undefined / empty body.
- removed `package.json`, `package-lock.json` and added them to `.gitignore`

YashKarthik commented:
pushed some new changes:
- Calls are forwarded directly, without hitting moderation/promptThreat endpoint, if content is empty/undefined. We return the provider's response (error) here.
- Replaced moderation logic in `forwarder.ts` with `Moderation` class.

colegottdank commented:
I'm closing because I had to cut a new branch off of this one to make changes. It has been merged.
